internal/net/httpserver.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package net
+
+import (
+	"errors"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/uber-go/atomic"
+)
+
+var (
+	errServerStopped    = errors.New("the server has been stopped")
+	errAlreadyListening = errors.New("the server is already listening")
+)
+
+// HTTPServer wraps an http.Server to listen asynchronously and allow stopping
+// it.
+type HTTPServer struct {
+	*http.Server
+
+	lock     sync.Mutex
+	listener net.Listener
+	done     chan error
+	stopped  atomic.Bool
+}
+
+// NewHTTPServer wraps the given http.Server into an HTTPServer.
+func NewHTTPServer(s *http.Server) *HTTPServer {
+	return &HTTPServer{
+		Server: s,
+		done:   make(chan error, 1),
+	}
+}
+
+// Listener returns the listener for this server or nil if the server isn't
+// yet listening.
+func (h *HTTPServer) Listener() net.Listener {
```

Comment:
True. I'll switch to RWLock and also drop the `atomic.Bool` to address @bombela's concerns.